core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/jndi/JndiDirectoryDAO.scala

```diff
@@ -0,0 +1,389 @@
+      if (!memberDns.isEmpty) {
+        val members = new BasicAttribute(Attr.member)
+        memberDns.foreach(subject => members.add(subject))
+        myAttrs.put(members)
+      }
+    }
+
+    def save(group: RawlsGroup): ReadWriteAction[RawlsGroup] = withContext { ctx =>
+      try {
+        val groupContext = new BaseDirContext {
+          override def getAttributes(name: String): Attributes = {
+            val myAttrs = new BasicAttributes(true) // Case ignore
+
+            val oc = new BasicAttribute("objectclass")
+            Seq("top", "workbenchGroup").foreach(oc.add)
+            myAttrs.put(oc)
+
+            myAttrs.put(new BasicAttribute(Attr.email, group.groupEmail.value))
+
+            addMemberAttributes(group.users, group.subGroups, myAttrs)
+
+            myAttrs
+          }
+        }
+
+        ctx.bind(groupDn(group.groupName), groupContext)
+
+      } catch {
+        case e: NameAlreadyBoundException =>
+          val myAttrs = new BasicAttributes(true) // Case ignore
+          addMemberAttributes(group.users, group.subGroups, myAttrs)
+          ctx.modifyAttributes(groupDn(group.groupName), DirContext.REPLACE_ATTRIBUTE, myAttrs)
+      }
+      group
+    }
+
+    def delete(groupRef: RawlsGroupRef): ReadWriteAction[Unit] = withContext { ctx =>
+      ctx.unbind(groupDn(groupRef.groupName))
+    }
+
+    def removeGroupMember(groupName: RawlsGroupName, removeMember: RawlsUserSubjectId): ReadWriteAction[Unit] = withContext { ctx =>
+      ctx.modifyAttributes(groupDn(groupName), DirContext.REMOVE_ATTRIBUTE, new BasicAttributes(Attr.member, userDn(removeMember)))
+    }
+
+    def removeGroupMember(groupName: RawlsGroupName, removeMember: RawlsGroupName): ReadWriteAction[Unit] = withContext { ctx =>
+      ctx.modifyAttributes(groupDn(groupName), DirContext.REMOVE_ATTRIBUTE, new BasicAttributes(Attr.member, groupDn(removeMember)))
+    }
+
+    def addGroupMember(groupName: RawlsGroupName, addMember: RawlsUserSubjectId): ReadWriteAction[Unit] = withContext { ctx =>
+      ctx.modifyAttributes(groupDn(groupName), DirContext.ADD_ATTRIBUTE, new BasicAttributes(Attr.member, userDn(addMember)))
+    }
+
+    def addGroupMember(groupName: RawlsGroupName, addMember: RawlsGroupName): ReadWriteAction[Unit] = withContext { ctx =>
+      ctx.modifyAttributes(groupDn(groupName), DirContext.ADD_ATTRIBUTE, new BasicAttributes(Attr.member, groupDn(addMember)))
+    }
+
+    def load(groupRef: RawlsGroupRef): ReadWriteAction[Option[RawlsGroup]] = withContext { ctx =>
+      Try {
+        val attributes = ctx.getAttributes(groupDn(groupRef.groupName))
+
+        Option(unmarshallGroup(attributes))
+
+      }.recover {
+        case e: NameNotFoundException => None
+
+      }.get
+    }
+
+    def load(groupRefs: TraversableOnce[RawlsGroupRef]): ReadWriteAction[TraversableOnce[RawlsGroup]] = withContext { ctx =>
```

Comment:
I'm not sure what this function should do if any of the groups are not found. Based on current callers, this is only called to fetch subgroups that were returned from a previous load. There is a very very small window during which a reference could be broken. It's very unlikely that it will happen but it could happen someday and would be nearly impossible to track down.

Can we detect groups that aren't found and warn-log them? I'd also like to see some scaladoc on this describing the behavior if a group is not found. (This isn't a problem for the singular `load` above which returns `Option[RawlsGroup]`.)